### PR TITLE
feat: update export command to export exam name and scores to csv

### LINKED ISSUE #250: Update: Get ExportCommand to work with Scores
Implementation Details:

1. Check the exams that are currently in the app. Add extra columns at the end of the export file for each exam
> image
> This should be the output if "Midterm" and "Final" are exams present in the application

2. Export scores for each person in the new columns.  If a person has no score associated to a particular exam, leave the cell blank.
> image
> This should be the output if Siti has records for both exams, Raj has records for none and Lucas has records for only midterm

### PR DESCRIPTION
**Pull request by danielleloh**
users should be able to export selected persons exam scores along with the other details stored when they use the export command.
fixes #250 

## Implementation Details
* when `export` command is called, the application will export details of all contacts shown with their exam scores as shown:
Screenshot 2024-03-31 at 3 54 01 PM

* column headings will be Exam:[EXAM_NAME] and values represent the scores of each person for that exam
* If there are no scores for that exam, the cell will be left blank.

---

add,
* read exams array from unfiltered json file (whole addressbook) to obtain the exam names as headings for the csv file. add "Exam:" to the front of the names so that headings will be displayed as Exam:[NAME_OF_EXAM] in csv file for ease of future exports. this also prevents errors if any of the exam names match the other fieldnames. (done in `buildCsvSchema` method).
* edit jsonTree for personsArray to change the nodes such that the examScores node and its children are removed and new key-value pairs [exam_name]: [score] is added for ease of writing jsonTree to csv schema built (done in`writeToCsvFile` method).

### CODE REVIEW

**Review by Zer0Legion — COMMENTED** [review]
Good abstraction!

File: src/main/java/seedu/address/commons/util/CsvUtil.java [hunk 1]
```diff
@@ -51,4 +61,73 @@ public static List<Map<String, String>> parseData(List<String[]> rows) {
...
+    public void writeToCsvFile(File csvFile, JsonNode personsArray, JsonNode examsArray) throws IOException {
+        CsvSchema csvSchema = buildCsvSchema(personsArray, examsArray);
+        CsvMapper csvMapper = new CsvMapper();
+
+        ArrayNode modifiedJsonTree = JsonNodeFactory.instance.arrayNode();
+        for (JsonNode person : personsArray) {
+            ObjectNode modifiedJsonNode = ((ObjectNode) person).deepCopy();
+            JsonNode examScores = modifiedJsonNode.get("examScores");
+            modifiedJsonNode.remove("examScores");
+            if (examScores != null && examScores.isArray()) {
```
Comment (jayllo-c) [inline src/main/java/seedu/address/commons/util/CsvUtil.java:117]: Do you think using SLAP here can improve the code?

File: src/main/java/seedu/address/commons/util/CsvUtil.java [hunk 2]
```diff
@@ -51,4 +61,73 @@ public static List<Map<String, String>> parseData(List<String[]> rows) {
...
     }
+
+    /**
+     * Builds the CSV schema from a JSON array.
+     *
+     * @param personsArray The JSON persons array to derive the schema from.
+     * @param examsArray The JSON exams array to derive the schema from.
+     * @return The CSV schema.
+     */
+    public CsvSchema buildCsvSchema(JsonNode personsArray, JsonNode examsArray) {
```
Comment (Pughal77) [inline src/main/java/seedu/address/commons/util/CsvUtil.java:72]: Nice Abstraction!

Comment (delishad21) [inline src/main/java/seedu/address/commons/util/CsvUtil.java:72]: Good that we abstracted all the csv related code to a utility class

File: src/main/java/seedu/address/commons/util/CsvUtil.java [hunk 3]
```diff
@@ -51,4 +61,73 @@ public static List<Map<String, String>> parseData(List<String[]> rows) {
...
+        firstObject.fieldNames().forEachRemaining(fieldName -> {
+            if (!fieldName.equals("examScores")) {
+                csvSchemaBuilder.addColumn(fieldName);
+                addedColumns.add(fieldName);
+            }
+        });
+
+        if (examsArray != null && examsArray.size() != 0) {
+            for (JsonNode exam : examsArray) {
+                String examName = "Exam:" + exam.get("name").asText();
```
Comment (delishad21) [inline src/main/java/seedu/address/commons/util/CsvUtil.java:86]: Good job implementing this!

**Review by Pughal77 — COMMENTED** [review]
Nice chnages. Lets work towards better code readabikity

**Review by delishad21 — COMMENTED** [review]
Well implemented overall!

**Review by jayllo-c — COMMENTED** [review]
Apart from the potential to improve slap, the feature looks well implemented

**Pull request merged**